### ISSUE TITLE
Advanced demo tools

### ISSUE THREAD

**Issue by kaadmy** [jdolan/quetoo#377]
One thing that I noticed after watching the greenlight trailer is Quetoo is lacking in demo tools, such as playback control and scripted camera movement for cinematic shots.

Easier said than done, but it _should_ be pretty simple to implement custom camera movement for spectators, server demos would be the tricky part.

**Comment by Paril** [comment]
Also serverrecord

**Comment by jdolan** [comment]
This is probably pretty important for 1.0. If people can make awesome Quetoo videos, more people will play Quetoo. Simple as that. Let's brainstorm how these camera controls could be implemented easily / minimally. Is there an ObjectivelyMVC UI component to this?

**Comment by Paril** [comment]
The controls aren't too big of a problem, but first iteration would probably just be basically spectator mode in demos + allowing you to play serverrecord demos in-game. Both of those are not allowed currently and would do wonders for what used to require a separate program (like KeyGrip) to edit demos.

After that, the harder part is rewind. Demos can go forward, but they can't go backwards currently, and that'll be pretty important for a scrubber.

**Comment by skies912** [comment]
What about using intermission ents as hard "camera" points spectators can swap to? Would provide extra POV for demo editing.

**Comment by jdolan** [comment]
Now that the entire UI in Quetoo uses ObjectivelyMVC, we have the pieces required to build a true first-class suite of demo tooling.

First, a home for demos visible to the current game module. I'm thinking a tab on the PlayViewController (currently houses Join, Create, & Player). Or possibly under the HomeViewController (currently Stats and Leaderboard, both of which are past-performance-looking, which demos are too).

The DemosViewController should use a single-select CollectionViewController to render available demos with:
 * Mapshot as Item background 
 * Map name
 * Date/time of recording
 * Length in mm::ss
 * Delete icon in top-right corner of cell

Because players may accumulate LOTS of demos, we may want to offer a Select for Item size. The default can match the MapListCollectionView, but having other options for showing 3, 4 or 5 columns will be handy. Perhaps this control, as well as a TextView for filtering demos by map name, can live just below the CollectionViewController in the accessoryView -- similar to the criteria options on JoinServerViewController.

Double clicking a demo begins playback of it. There can also be a Play button in the accessoryView that does the same for the selected Item.

For demo playback, we need to augment or overlay atop the HUD, I think? Or perhaps this is just a new subview in the HUD, like the scoreboard, that is hidden most of the time.

We need to show playback controls (pause, fast forward, rewind, perhaps a scrubber). We also want to let users "spectate" (free fly around) their demo, or anchor their view to the POV of the recording. This should be toggled with keystrokes, and there should be an on-screen indicator (vector icon) showing which state you're in.

I'm fuzzy on the specifics for the HUD overlay / in-demo playback controls. I'll have Claude create some mockups atop the existing HUDs.

**Comment by jdolan** [comment]
Default HUD, for mockup purposes:

<img width="3840" height="2160" alt="Image" src="https://github.com/user-attachments/assets/efe86208-0d9d-4bcd-88b2-f2e16201b1a5" />

Chrome HUD, for mockup purposes:

<img width="3456" height="2104" alt="Image" src="https://github.com/user-attachments/assets/99f687a2-5162-4b7a-a7e2-4530400f89b6" />

**Comment by jdolan** [comment]
# Demo tooling overhaul (quetoo#377)

## Context

Issue #377 asks for first-class demo tooling: a demo browser in the ObjectivelyMVC UI, in-HUD
playback controls (pause/scrub/rewind), and a spectator/free-fly vs. POV-locked toggle during
playback. The blocker, identified while reviewing the issue, is architectural: the `.demo` file
is a flat, unindexed stream of raw net messages (`Cl_WriteDemoMessage`, `cl_demo.c:96`), read
strictly forward one frame at a time by `Sv_GetDemoMessage` (`sv_send.c:329`) with no seek
support. There's even a standing FIXME on that function: *"We need a mechanism to indicate frame
completion, or we need a timecode in our demos."* Rewind, scrubbing, and cheap demo-list metadata
(map name, duration, thumbnail) all require a format that supports seeking to an arbitrary point
without replaying from byte 0.

Decision (confirmed by Jay): no backward compatibility with existing `.demo` files — nobody has
enough of them recorded for it to matter, so the format can change freely.

The design that emerged requires far less new machinery than first thought, because Quetoo already
has a client-driven mechanism for requesting a full, non-delta ("uncompressed") frame.
`Cl_WriteMovementCommand` (`cl_cmd.c:69-79`) reports the client's last-received frame number to the
server on every move command; while recording and the demo file is still empty
(`cls.demo.file && Fs_Tell(cls.demo.file) == 0`), it deliberately lies and sends `-1`. The server
takes this at face value (`sv_client.c:369`, `cl->last_frame = last_frame`), and
`Sv_WriteClientFrame` (`sv_entity.c:115`) treats `last_frame < 0` as "client is asking for a
retransmit," sending a full uncompressed frame in response. This is exactly how `Cl_Record_f`
already gets its initial header data today, not a wait-and-hope. `last_frame`'s doc comment
(`sv_types.h:267`, "Last acknowledged frame number for delta compression") confirms it exists
solely to drive delta compression, so lying about it repeatedly has no gameplay-visible side
effect beyond the full-frame resync itself.

Periodic keyframes can reuse this exact path: extend the client's condition in
`Cl_WriteMovementCommand` to also send `-1` whenever it's time for the next scheduled resync while
recording, not only at the very start. This needs no server-side changes at all: no new userinfo
cvar, no new `sv_client_t` bookkeeping, no new branch in `Sv_WriteClientFrame`. The server already
does the right thing whenever a client's reported `last_frame` is negative, regardless of why.

Bandwidth check (calculated from the wire-encoding code, not empirically benchmarked — see
verification section): a fully uncompressed frame for a busy 32-player scene with projectiles is
roughly ~5.5KB (`Net_WriteDeltaEntity` full-entity cost ≈53B × ~100 live entities, plus a full
`Net_WriteDeltaPlayerState` ≈160B). This is sent **only to the recording client**, never broadcast,
so server/other-player cost is negligible. For the recording client, a 1/sec interval adds ~5.5KB/s
sustained (a large fraction of their normal traffic) and forces a visible `no_lerp` interpolation
pop (`cl_entity.c:346`) roughly once a second. A 5s interval amortizes to ~1.1KB/s and one barely
perceptible pop every 5s. Recommend hardcoding 5s rather than exposing a cvar — this is an
engine-internal tuning knob, not something a player needs to touch, and an extra cvar is a
support/discoverability hazard more than it's useful.

## Phase 1 — Demo format v2: recording-side authoring

**Goal:** client-recorded `.demo` files gain periodic full-state resync points and a header index,
with zero broadcast cost to non-recording players.

**Client only (`src/client`) — no server changes needed:**
- A hardcoded interval constant, e.g. `#define DEMO_KEYFRAME_INTERVAL 5000` (ms) near the other
  demo constants in `cl_demo.c`, plus a `cls.demo` field tracking the next scheduled resync time.
  Not a cvar — this is an internal tuning value, not something a player needs to configure.
- `Cl_WriteMovementCommand` (`cl_cmd.c:69-79`): extend the existing condition —
  `if (!cl.frame.valid || (cls.demo.file && (Fs_Tell(cls.demo.file) == 0 || cl.time >=
  cls.demo.next_keyframe_time))) { Net_WriteLong(buf, -1); ...schedule next... }` — so the client
  requests a fresh uncompressed frame periodically while recording, the same way it already does
  once at the very start. This is the only wire-visible change in this phase, and it's entirely
  request-driven from the existing client, using an existing server response path.
- `Cl_Record_f` (`cl_demo.c:142`) resets `cls.demo.next_keyframe_time` when recording starts.
- `Cl_WriteDemoMessage` (`cl_demo.c:96`) already detects an uncompressed frame today via
  `cl.frame.delta_frame_num < 0` (`cl_demo.c:103`, `cl_entity.c:277-283`) to trigger the initial
  header write. Extend this: every subsequent occurrence (i.e. a periodic resync this client just
  requested) gets appended as an entry — `(cl.frame.frame_num, Fs_Tell(cls.demo.file))` — to a
  growing in-memory keyframe list on `cls.demo`. No file re-scanning is needed anywhere; the index
  is built incrementally, in real time, as the demo is recorded.
- Each raw message write also gets a small fixed-size prefix carrying `cl.frame.frame_num` (client
  already knows this — `cl.frame.frame_num`, set in `cl_entity.c:277`), in addition to the existing
  length prefix. This keeps `Sv_GetDemoMessage`/the future demo-serving code "blob agnostic" (it
  doesn't need a protocol parser to find frame boundaries) while still letting it locate an
  arbitrary timecode by peeking prefixes during a forward scan from a keyframe — directly resolving
  the FIXME on `sv_send.c:325-327`.
- Replace the current ad hoc write-out with a fixed-size header at byte offset 0: magic + version
  (so future format changes can version-check; old files are simply rejected), map name
  (`cl.config_strings[CS_BSP]`, `shared.h:41` — known immediately at record start), a `favorite`
  bool (starts `false`; toggled later in place from the UI, see Phase 3 — a single-field in-place
  rewrite, safe on a closed file), `thumbnail_length`, `data_offset` (byte offset of the first real
  packet-data byte, i.e. right after the embedded thumbnail), and three fields left as placeholders
  until `Cl_Stop_f`: `duration`, `keyframe_count`, `keyframe_table_offset`. Fixed-size means the
  browser (Phase 3) can always read it with one `Fs_Read` at offset 0 — no scanning, no
  trailing-footer convention to get right.
- Thumbnail, captured at `Cl_Record_f` (`cl_demo.c:142`) time, embedded directly in the file rather
  than reusing per-map mapshots: dispatch `R_Screenshot` (`r_image.h:27`, `r_image.c:156`) to grab
  the currently rendered frame, downsample to a fixed height of 480px, JPEG-encode it — reusing the
  same `IMG_SaveJPG_IO` call `Img_WriteJPG` already uses (`image.c:314-333`), but backed by an
  SDL in-memory IO stream instead of `SDL_IOFromFile`, so the encoded bytes land in a buffer instead
  of a separate file. `R_Screenshot`'s existing capture path offloads JPEG encoding to a background
  thread (`r_image.c:166`, `Thread_Create(..., THREAD_NO_WAIT)`) because full-resolution screenshots
  are relatively expensive; a 480px-tall thumbnail is small enough to encode synchronously at record
  start instead, which keeps the format simple (header + thumbnail must be fully sized before frame
  data can begin at a known `data_offset`) — worth confirming at implementation time that a
  synchronous encode of this size is fast enough not to hitch the frame `record` was issued on.
  The thumbnail bytes are written immediately after the fixed header, before any frame/packet data.
- At `Cl_Stop_f` (`cl_demo.c:121`): the in-memory keyframe list built during recording gets appended
  to the file at the current (end) position — this is `keyframe_table_offset`. Then `Fs_Seek` back
  to offset 0 and patch the placeholder fields (`duration = cl.frame.frame_num *
  QUETOO_TICK_MILLIS`, `keyframe_count`, `keyframe_table_offset`) into the fixed header before
  closing. No footer/EOF convention anywhere in the format.
- The resulting ~5.5KB worst-case full frame already fragments correctly via the existing
  `Sv_SendClientDatagram` overflow logic (`sv_send.c:265`) — no changes needed there.

**Known limitation to flag (not fixed by this plan):** client-side recording only captures what
that player's own PVS/PHS-limited netchannel received — the same visibility limit any demo format
has always had. Free-fly camera during playback (Phase 4) can only see what was actually
transmitted to the recording player at record time. True unlimited "director cam" would need
server-side recording (the "serverrecord" idea from the issue's first comment) — out of scope
here, worth its own follow-up issue if wanted.

## Phase 2 — Server-side seek (playback)

**Goal:** `demo <name>` (`Sv_Demo_f`, `sv_admin.c:90`) playback can jump to an arbitrary timecode
instead of only playing forward from byte 0.

- On `SV_ACTIVE_DEMO` init, read the new fixed-size header once (one `Fs_Read` at offset 0) to get
  `keyframe_table_offset`/`keyframe_count`, then read the keyframe table itself, instead of playing
  byte 0 forward blind.
- Add a seek entry point (e.g. `Sv_SeekDemo(ms)`): pick the keyframe at/before the target from the
  index, `Fs_Seek` the demo file to its offset, then fast-scan forward via `Sv_GetDemoMessage`
  (peeking the new per-message frame_num prefix, not decoding message contents) discarding
  messages without transmitting until the target frame_num is reached, then resume normal
  transmission via the existing loop in `Sv_SendClientPackets` (`sv_send.c:366-406`).
- Needs a client→server command to trigger a seek when the user drags the scrubber (Phase 4) — this
  should follow whatever existing pattern Quetoo uses for reliable client→server console commands
  (the same family as `record`/`stop`/`say`); the exact dispatch point needs a short follow-up look
  at `sv_user.c`/`sv_ccmds.c` when this phase starts, not fully nailed down yet.
- Pause: simplest implementation is client-side — stop advancing local prediction/rendering time
  while still acking, OR a server-side "don't call `Sv_GetDemoMessage` this tick" latch keyed off a
  client→server pause command. Needs a design call at implementation time; either is small.

## Phase 3 — DemosViewController (browser UI)

**Goal:** a CollectionView-based demo browser, following the exact shape of the existing map
browser.

- Recommend placing this under `HomeViewController` (`src/cgame/common/ui/home/HomeViewController.c:62-76`)
  alongside Stats/Leaderboard, not `PlayViewController` — Jay's own framing in the issue was that
  Stats/Leaderboard/Demos are all "past-performance-looking," and `HomeViewController`'s tab-wiring
  is the identical 3-line `addChildViewController` pattern as Play's, so the choice doesn't add
  complexity either way.
- New `DemosViewController`/`DemosCollectionView`/`DemosCollectionItemView`, directly mirroring
  `MapListCollectionView`/`MapListCollectionItemView` (`src/cgame/common/ui/play/`): async
  population via `cgi.Thread` + `cgi.EnumerateFiles("demos/*.demo", ...)` (matching
  `MapListCollectionView.c:112-218,289`), one item per demo.
- Per-item data (map name, date, duration) now comes directly from each demo's fixed-size header
  (Phase 1) — one `Fs_Read` at offset 0 per file, not a full scan.
- Thumbnail comes directly from the embedded JPG in each demo's header (Phase 1): read
  `thumbnail_length` bytes starting right after the fixed header, decode via the same raster-image
  path `MapListCollectionView.c` already uses for its fallback case, no per-demo sidecar file needed.
- Star/favorite: item view gets a toggle that flips the `favorite` bool in place in that one demo's
  header (single-field rewrite, per Phase 1); the accessoryView filter controls (alongside item-size
  and map-name filter, per the issue) gain a "favorites only" option.
- Item size selector / map-name filter TextView, per the issue: same accessoryView pattern already
  used by `JoinServerViewController`'s criteria controls.
- Scrubber control: ObjectivelyMVC's `Slider` (`Sources/ObjectivelyMVC/Slider.h`) — `initWithFrame`,
  `setValue`, `SliderDelegate.didSetValue` — is a direct fit, no new control needed.

## Phase 4 — HUD demo-playback-controls overlay

**Goal:** pause/fast-forward/rewind/scrub controls and a spectate-vs-POV-lock toggle, visible only
during demo playback.

- Demo-playback state is already client-visible today: `cgi.client->demo_server`
  (`cl_types.h:419-422`, "True if the client is viewing a demo") — already used by
  `cg_predict.c:34` and `LoadingViewController.c:103`. A new `DemoControlsView` gates its
  visibility on this exact flag, the same way `HudViewController` already gates the scoreboard on
  `STAT_SCORES` and intermission on `cg_state.next_map.active` (`HudViewController.c:401-421`).
- Wiring follows the existing scoreboard/intermission pattern exactly: alloc + add in
  `loadView`/`reload` (`HudViewController.c:103-119,279-333`), visibility set each frame in
  `updateWithFrame` (`:386-437`), which is invoked every frame from `cg_hud.c:65-66`.
- POV-lock vs. free-fly is not new infrastructure — `G_ClientChaseThink`/`Next`/`Previous`
  (`g_client_chase.c:27,80,110`) already implement exactly this, just not wired to a demo-specific
  keybind/toggle. `SpectatorView.c:42-51` already renders a text indicator off
  `STAT_SPECTATOR`/`STAT_CHASE` — reuse it, or extend it with a vector icon.
- Icons: no existing vector-icon toggle precedent (the current spectate indicator is text-only) —
  follow the general asset convention (`Cg_LoadImageScaled`, SVG-first, `cg_ui.c:34-64`, atlas-cached
  via `HudViewController::image`, `:130-162`) for new pause/play/rewind/POV icons.

## Sequencing

Phase 1 is the prerequisite for everything else and should land first. Phase 2 (seek) and Phase 3
(browser UI) can proceed in parallel once Phase 1 lands, since the browser only needs the header
format and the seek command only needs the keyframe index — neither depends on the other. Phase 4
(HUD overlay) depends on Phase 2 existing (no point building scrub/rewind UI with nothing to drive).

## Verification

- Phase 1: record a demo against a local dedicated server with bots (`+game <mod> +map <map>` per
  existing local-testing conventions), inspect the resulting `.demo` file's header and appended
  keyframe table with a hex dump to confirm the patched-back fields (duration, keyframe_count,
  keyframe_table_offset) and keyframe entries are correct; confirm non-recording bots/players see
  no behavior change (delta frames unaffected).
- Bandwidth: before committing to the 5s default, optionally run the calculated worst case for
  real — a bot-filled dedicated server (32 bots, weapon spam for projectile density) recording a
  demo at 1s vs 5s intervals, comparing actual captured packet sizes against the ~5.5KB estimate.
- Phase 2: play back a recorded demo, issue a seek command mid-playback, confirm the client lands
  at the correct point without visual corruption (missing entities, stale player state).
- Phase 3: launch client, open the new Demos tab, confirm list population, thumbnail (once
  implemented), filter, and double-click-to-play all work.
- Phase 4: play back a demo, confirm playback controls appear only during playback (not live play),
  confirm scrub/rewind lands correctly, confirm spectate/POV toggle switches between
  `G_ClientChaseThink`-driven POV lock and free-fly.

**Comment by jdolan** [comment]
Fixed im #1049
